---
title: "2025-02-19_Rotation_GSEA_KD"
author: "wisdom"
date: "2025-02-19"
output: html_document
---

### 1. Install and Load Required Packages
```{r}
if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install(version = "3.20")
BiocManager::install(c("clusterProfiler", "org.Hs.eg.db", "org.Mm.eg.db", "org.Rn.eg.db"))
install.packages("readxl")
install.packages("here")
install.packages("gitcreds")
install.packages("biomaRt")
install.packages("cowplot")
```
Load Libraries
```{r}
library(clusterProfiler)
library(tidyr)
library(readxl)
library(here)
library(gitcreds)
library(org.Hs.eg.db)
library(org.Mm.eg.db)
library(org.Rn.eg.db)
library(biomaRt)
library(cowplot)
```

### 2. Prepare ranked gene list

#### 2.1 Load differerially genes  between **core** and **rim** in **mouse** and **rat** after ANGPL7 knockdown. Supporting Dataset S03<https://www.pnas.org/doi/suppl/10.1073/pnas.2214888120/suppl_file/pnas.2214888120.sd03.xlsx> from paper <https://www.pnas.org/doi/10.1073/pnas.2214888120>
```{r}
# Load mouse and rat genes
mm_core_KDvsNT4 <- read_excel(here("data","pnas.2214888120.sd03.xlsx"), sheet = "TPM_mm_core_KDvsNT4", range =cell_cols(c("A","I")))
rn_core_KDvsNT4 <- read_excel(here("data","pnas.2214888120.sd03.xlsx"), sheet = "TPM_rn_coreKDvsNT4", range =cell_cols(c("A","I")))
mm_rim_KDvsNT4 <- read_excel(here("data","pnas.2214888120.sd03.xlsx"), sheet = "TPM_mm_rim-KDNT4", range =cell_cols(c("A","I")))
rn_rim_KDvsNT4 <- read_excel(here("data","pnas.2214888120.sd03.xlsx"), sheet = "TPM_rn_rim-KDvsNT4", range =cell_cols(c("A","I")))
```

#### 2.2 Rank gene list
```{r}
mm_core_KDvsNT4_vector <- sort(setNames(mm_core_KDvsNT4$logFC, mm_core_KDvsNT4$ensembl_gene_id), decreasing = TRUE)
rn_core_KDvsNT4_vector <- sort(setNames(rn_core_KDvsNT4$logFC, rn_core_KDvsNT4$ensembl_gene_id), decreasing = TRUE)
mm_rim_KDvsNT4_vector <- sort(setNames(mm_rim_KDvsNT4$logFC, mm_rim_KDvsNT4$ensembl_gene_id), decreasing = TRUE)
rn_rim_KDvsNT4_vector <- sort(setNames(rn_rim_KDvsNT4$logFC, rn_rim_KDvsNT4$ensembl_gene_id), decreasing = TRUE)
```

### 3 Prepare ranked genes with both fc and pval
```{r}
mm_core_KDvsNT4_vector_p <- sign(mm_core_KDvsNT4$logFC)*(-log10(mm_core_KDvsNT4$P.Value)) # we will use the signed p values from spatial DGE as ranking
names(mm_core_KDvsNT4_vector_p) <- mm_core_KDvsNT4$ensembl_gene_id # genes as names#
mm_core_KDvsNT4_vector_p <- sort(mm_core_KDvsNT4_vector_p, decreasing = TRUE) # sort genes by ranking

rat_core_vs_rim_vector_p <- sign(rat_core_vs_rim$logFC)*(-log10(rat_core_vs_rim$P.Value)) # we will use the signed p values from spatial DGE as ranking
names(rat_core_vs_rim_vector_p) <- rat_core_vs_rim$ensembl_gene_id # genes as names#
rat_core_vs_rim_vector_p <- sort(rat_core_vs_rim_vector_p, decreasing = TRUE) # sort genes by ranking

```

### 4 Prepare ranked genes with t statistic
```{r}
mm_core_KDvsNT4_vector_t <- sort(setNames(mm_core_KDvsNT4$t, mm_core_KDvsNT4$ensembl_gene_id), decreasing = TRUE)
rn_core_KDvsNT4_vector_t <- sort(setNames(rn_core_KDvsNT4$t, rn_core_KDvsNT4$ensembl_gene_id), decreasing = TRUE)
mm_rim_KDvsNT4_vector_t <- sort(setNames(mm_rim_KDvsNT4$t, mm_rim_KDvsNT4$ensembl_gene_id), decreasing = TRUE)
rn_rim_KDvsNT4_vector_t <- sort(setNames(rn_rim_KDvsNT4$t, rn_rim_KDvsNT4$ensembl_gene_id), decreasing = TRUE)
```

### 3. Prepare VI clusters as genesets

#### 3.1 Load VI clusters 
```{r}
# Load VI gene clusters
VI_gene_cluster <- readRDS(here("data","VI_gene_clusters.rds"))
```

```{r}
VI_gene_cluster$gene_symbol <- rownames(VI_gene_cluster)
```

#### 3.2 Convert Gene symbolls to Entrez IDs
```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")
VI_gene_cluster$ensemblID = mapIds(org.Hs.eg.db,
                    keys=VI_gene_cluster$gene_symbol, 
                    column="ENSEMBL",
                    keytype="SYMBOL",
                    multiVals="first")

#Manual Annotation
VI_gene_cluster[VI_gene_cluster$gene_symbol == "IGHV3.73","ensemblID"] <- "ENSG00000211976"
VI_gene_cluster[VI_gene_cluster$gene_symbol == "DGCR5.1","ensemblID"] <-  "ENSG00000273032"
VI_gene_cluster[VI_gene_cluster$gene_symbol == "ARNTL2","ensemblID"] <- "ENSG00000029153"
VI_gene_cluster[VI_gene_cluster$gene_symbol == "WT1.AS","ensemblID"] <- "ENSG00000183242"
VI_gene_cluster[VI_gene_cluster$gene_symbol == "H1.5","ensemblID"] <- "ENSG00000184357"
VI_gene_cluster[VI_gene_cluster$gene_symbol == "CASTOR3","ensemblID"] <- "ENSG00000239521"
VI_gene_cluster[VI_gene_cluster$gene_symbol == "GAPDHP1","ensemblID"] <- "ENSG00000228232"
VI_gene_cluster[VI_gene_cluster$gene_symbol == "AC010329.5","ensemblID"] <- "ENSG00000280171"
VI_gene_cluster[VI_gene_cluster$gene_symbol == "MIR29B2CHG","ensemblID"] <- "ENSG00000203709"
VI_gene_cluster[VI_gene_cluster$gene_symbol == "AC012651.1","ensemblID"] <- "ENSG00000258461" 
VI_gene_cluster[VI_gene_cluster$gene_symbol == "AL663070.1","ensemblID"] <- "ENSG00000168389"

# Rename Clusters
VI_gene_cluster[VI_gene_cluster$Cluster == "1","Cluster"] <- "Cluster1"
VI_gene_cluster[VI_gene_cluster$Cluster == "2","Cluster"] <- "Cluster2"
VI_gene_cluster[VI_gene_cluster$Cluster == "3","Cluster"] <- "Cluster3"
VI_gene_cluster[VI_gene_cluster$Cluster == "4","Cluster"] <- "Cluster4"
```


#### 3.2 Convert VI clusters from human genes to mouse and rat orthologs
```{r}
mm_geneset_cluster1 <- getHomologs(VI_gene_cluster$ensemblID, species_from = "homo_sapiens", species_to = "mus_musculus" )


rat_geneset_cluster1 <- getHomologs(VI_gene_cluster$ensemblID, species_from = "homo_sapiens", species_to = "rattus_norvegicus" )
```


```{r}
colnames(mm_geneset_cluster1)[1] <- "ensemblID" 
colnames(rat_geneset_cluster1)[1] <- "ensemblID" 
mouse_geneset <- merge(VI_gene_cluster,mm_geneset_cluster1, by="ensemblID")
rat_geneset <- merge(VI_gene_cluster,rat_geneset_cluster1, by="ensemblID")
#Remove extra columns
mouse_geneset$ensemblID <- NULL
mouse_geneset$gene_symbol <- NULL
rat_geneset$ensemblID <- NULL
rat_geneset$gene_symbol <- NULL
#Rename cloumns
colnames(mouse_geneset)[1] <- "TermID" 
colnames(mouse_geneset)[2] <- "geneID"
colnames(rat_geneset)[1] <- "TermID" 
colnames(rat_geneset)[2] <- "geneID"
```

### 4. Perform GSEA
```{r}
#Mouse Geneset & mouse Genelist from core and rim
gsea_results_mouse_rim <- GSEA(mm_core_KDvsNT4_vector_t, TERM2GENE = mouse_geneset, eps= 1e-300)
gsea_results_mouse_core <- GSEA(mm_rim_KDvsNT4_vector_t, TERM2GENE = mouse_geneset, eps= 1e-300)

#Rat Geneset & Rat GEnelist
gsea_results_rat_rim <- GSEA(rn_core_KDvsNT4_vector_t, TERM2GENE = rat_geneset, eps= 1e-300)
gsea_results_rat_core <- GSEA(rn_rim_KDvsNT4_vector_t, TERM2GENE = rat_geneset, eps= 1e-300)
```
#### 5. Plot GSEA Plots
```{r}
# Mouse Core Plots
mouse_core_cluster1_plot <- gseaplot(gsea_results_mouse_core, geneSetID = "Cluster1", title = "Mouse-Core-Cluster1")
mouse_core_cluster2_plot <- gseaplot(gsea_results_mouse_core, geneSetID = "Cluster2", title = "Mouse-Core-Cluster2")
mouse_core_cluster3_plot <- gseaplot(gsea_results_mouse_core, geneSetID = "Cluster3", title = "Mouse-Core-Cluster3")
mouse_core_cluster4_plot <- gseaplot(gsea_results_mouse_core, geneSetID = "Cluster4", title = "Mouse-Core-Cluster4")

# Mouse Rim Plots
mouse_rim_cluster1_plot <- gseaplot(gsea_results_mouse_rim, geneSetID = "Cluster1", title = "Mouse-Rim-Cluster1")
mouse_rim_cluster2_plot <- gseaplot(gsea_results_mouse_rim, geneSetID = "Cluster2", title = "Mouse-Rim-Cluster2")
mouse_rim_cluster3_plot <- gseaplot(gsea_results_mouse_rim, geneSetID = "Cluster3", title = "Mouse-Rim-Cluster3")
mouse_rim_cluster4_plot <- gseaplot(gsea_results_mouse_rim, geneSetID = "Cluster4", title = "Mouse-Rim-Cluster4")

# Rat Core Plots
rat_core_cluster1_plot <- gseaplot(gsea_results_rat_core, geneSetID = "Cluster1", title = "Rat-Core-Cluster1")
rat_core_cluster2_plot <- gseaplot(gsea_results_rat_core, geneSetID = "Cluster2", title = "Rat-Core-Cluster2")
rat_core_cluster3_plot <- gseaplot(gsea_results_rat_core, geneSetID = "Cluster3", title = "Rat-Core-Cluster3")
rat_core_cluster4_plot <- gseaplot(gsea_results_rat_core, geneSetID = "Cluster4", title = "Rat-Core-Cluster4")

# Mouse Rim Plots
rat_rim_cluster1_plot <- gseaplot(gsea_results_rat_rim, geneSetID = "Cluster1", title = "Rat-Rim-Cluster1")
rat_rim_cluster2_plot <- gseaplot(gsea_results_rat_rim, geneSetID = "Cluster2", title = "Rat-Rim-Cluster2")
rat_rim_cluster3_plot <- gseaplot(gsea_results_rat_rim, geneSetID = "Cluster3", title = "Rat-Rim-Cluster3")
rat_rim_cluster4_plot <- gseaplot(gsea_results_rat_rim, geneSetID = "Cluster4", title = "Rat-Rim-Cluster4")
```